### DISCUSSION TITLE
Love this software. How can I use it to buy 1 DAI on quickswap.exchange?

### DISCUSSION THREAD

**Discussion by 621625** [gochain/web3#216] [General]
Hi,

I love gochain/web3. It rocks. You rock.

How can I use it to buy 1 DAI on quickswap.exchange? This might be possible using py.web3. But I am also testing gochain/web3 now and so far I prefer this software.

Ok, so we can already deploy and interact using goweb3. But concretely, let's say I want to buy 1 DAI on quickswap.exchange? Can I do this with web3? Any examples about how to interact with DEX using goweb3?

**Comment by treeder** [comment]
You'd have to find the abi files for quickswap if they are available, then you can use that with the `--abi` flag and then use `web3 contract call`

**Reply by 621625** [reply]
Ok, I have saved the abi in a file called qs.abi. I also know the name of the function to call, it's this:

swapExactETHForTokens(uint256 amountOutMin, address[] path, address to, uint256 deadline)

Now I am trying to figure out how the exact command should look like. Any advice would be appreciated.

**Reply by 621625** [reply]
I think it should be something like this:

web3 contract call --address                                0x92c819e24Cb020e35Fce434a33E39Fd5C85AB2aA --abi          qs.abi --function swapExactETHForTokens

But I still need to know how to finalize this input. What else do I have to add to make this command swap 1 MATIC for DAI? Exactly how should the command look?

**Reply by 621625** [reply]
ERROR: Error calling contract: mismatched argument (4) and parameter (0) counts

**Reply by 621625** [reply]
This is the transaction that I"m trying to replicate:

https://polygonscan.com/tx/0xcaf7358788aed4fd73b4fc26717e44a54750811fc1c62e6997db9bb0ae9f7585

**Reply by treeder** [reply]
So just try adding your params to the end of the that command, like:

```sh
web3 contract call --address 0x92c819e24Cb020e35Fce434a33E39Fd5C85AB2aA --abi qs.abi --function swapExactETHForTokens A B C D
```

**Reply by 621625** [reply]
This is the function including parameters:

swapExactETHForTokens(uint256__amountOutMin, address[] path, address to,uint256 deadline

This is what I have tried:

--function swapExactETHForTokens 0.1 ContractAdress MyWalletAddress 69

I'm not sure about the syntax and about the values of the parameters. Now I assumed I can just put the parameters like that without any commas or special characters between them. Is this the correct syntax?

And regarding the parameters one by one:

uint256__amountOutMin: Should this be the amount to be swapped written like so: 0.1?
address[] path: Should this be the address of the contract?
address to: Should this be my wallet address?
uint256 deadline: Should this be the number 69 or 42? Or something else?

If I can get the hang of it I can start doing this myself but I would need an example or two like this to get started.

**Reply by treeder** [reply]
You should ask the people that wrote the contract, but a quick google found this: https://docs.uniswap.org/protocol/V2/reference/smart-contracts/router-02#swapexactethfortokens

**Reply by 621625** [reply]
Ok, now I have tried using the input data from the corresponding parameters I found. So the command looks like this:

web3 contract call --address 0x92c819e24Cb020e35
Fce434a33E39Fd5C85AB2aA --abi qs.abi --function
swapExactETHForTokens 138607273911012674 0d500b1
d8e8ef31e21c99d1db9a6444d3adf12708f3cf7ad23cd3ca
dbd9735aff958023239c6a063 e5d77ab5fea649ca621f4f
6c8a9b542fef88ab8b 3260059519

It generates this error message and I don't know why. I just want to make exactly the same transaction as I already have done but using the CLI and web3 this time. Error message:

ERROR: Error calling contract: unsupported input type 4

**Reply by 621625** [reply]
There is an error above, the path should actually have two addresses.

**Reply by 621625** [reply]
I have updated the command to separate the two addresses for the parameter. Now it looks like this:

web3 contract call --address 0x92c819e24Cb020e35Fce434a33E39Fd5C85AB2aA --abi qs.abi --function swapExactETHForTokens 138607273911012674 0d500b1d8e8ef31e21c99d1db9a6444d3adf1270 8f3cf7ad23cd3cadbd9735aff958023239c6a063 e5d77ab5fea649ca621f4f6c8a9b542fef88ab8b 3260059519

However, this generates the following error message:

ERROR: Error calling contract: mismatched argument (4) and parameter (5) counts

Is my syntax correct or do I need to add some parenthesis or separators to the command?

**Reply by treeder** [reply]
I'm not sure this can handle an array of inputs like that `path` parameter requires, that's probably the issue.

**Comment by 621625** [comment]
Yes, I thought the same thing yesterday. Maybe you haven't implemented arrays for parameters. Please do. I assume the syntax is correct then, just plain text without [ or ". Hope you implement this, it would be a very cool feature.

-------- Original Message --------

**Comment by treeder** [comment]
Related issue: https://github.com/gochain/web3/issues/218
